// @deskthing/types/dist/clients/clientData.js
export var ClientConnectionMethod;
(function (ClientConnectionMethod) {
    ClientConnectionMethod[ClientConnectionMethod["Unknown"] = 0] = "Unknown";
    ClientConnectionMethod[ClientConnectionMethod["LAN"] = 1] = "LAN";
    ClientConnectionMethod[ClientConnectionMethod["Localhost"] = 2] = "Localhost";
    ClientConnectionMethod[ClientConnectionMethod["ADB"] = 3] = "ADB";
    ClientConnectionMethod[ClientConnectionMethod["NDIS"] = 4] = "NDIS";
    ClientConnectionMethod[ClientConnectionMethod["Bluetooth"] = 5] = "Bluetooth";
    ClientConnectionMethod[ClientConnectionMethod["Internet"] = 6] = "Internet";
})(ClientConnectionMethod || (ClientConnectionMethod = {}));
export var PlatformIDs;
(function (PlatformIDs) {
    PlatformIDs["ADB"] = "adb";
    PlatformIDs["WEBSOCKET"] = "websocket";
    PlatformIDs["BLUETOOTH"] = "bluetooth";
    PlatformIDs["MAIN"] = "main";
})(PlatformIDs || (PlatformIDs = {}));
export var ClientPlatformIDs;
(function (ClientPlatformIDs) {
    ClientPlatformIDs[ClientPlatformIDs["Unknown"] = 0] = "Unknown";
    ClientPlatformIDs[ClientPlatformIDs["Desktop"] = 1] = "Desktop";
    ClientPlatformIDs[ClientPlatformIDs["Tablet"] = 2] = "Tablet";
    ClientPlatformIDs[ClientPlatformIDs["Iphone"] = 3] = "Iphone";
    ClientPlatformIDs[ClientPlatformIDs["CarThing"] = 4] = "CarThing";
})(ClientPlatformIDs || (ClientPlatformIDs = {}));
export var ConnectionState;
(function (ConnectionState) {
    ConnectionState[ConnectionState["Connected"] = 0] = "Connected";
    ConnectionState[ConnectionState["Established"] = 1] = "Established";
    ConnectionState[ConnectionState["Connecting"] = 2] = "Connecting";
    ConnectionState[ConnectionState["Disconnecting"] = 3] = "Disconnecting";
    ConnectionState[ConnectionState["Disconnected"] = 4] = "Disconnected";
    ConnectionState[ConnectionState["Failed"] = 5] = "Failed";
})(ConnectionState || (ConnectionState = {}));
export var ProviderCapabilities;
(function (ProviderCapabilities) {
    ProviderCapabilities[ProviderCapabilities["CONFIGURE"] = 0] = "CONFIGURE";
    ProviderCapabilities[ProviderCapabilities["PING"] = 1] = "PING";
    ProviderCapabilities[ProviderCapabilities["COMMUNICATE"] = 2] = "COMMUNICATE";
})(ProviderCapabilities || (ProviderCapabilities = {}));
/**
 * @module deskthing/client
 * @description
 * View mode enum used for configuring display states
 * @see {@link ClientPreferences.appTrayState}
 * @see {@link MiniplayerSettings.state}
 */
export var ViewMode;
(function (ViewMode) {
    ViewMode["HIDDEN"] = "hidden";
    ViewMode["PEEK"] = "peek";
    ViewMode["FULL"] = "full";
})(ViewMode || (ViewMode = {}));
/**
 * @module deskthing/client
 * @description
 * Volume mode enum used for configuring volume control behavior
 * @see {@link ClientPreferences.volume}
 */
export var VolMode;
(function (VolMode) {
    VolMode["WHEEL"] = "wheel";
    VolMode["SLIDER"] = "slider";
    VolMode["BAR"] = "bar";
})(VolMode || (VolMode = {}));
