// @deskthing/types/dist/deskthing/mappings.js
/**
 * @deprecated - use {@link EventMode} instead
 */
export var EventFlavor;
(function (EventFlavor) {
    EventFlavor[EventFlavor["KeyUp"] = 0] = "KeyUp";
    EventFlavor[EventFlavor["KeyDown"] = 1] = "KeyDown";
    EventFlavor[EventFlavor["ScrollUp"] = 2] = "ScrollUp";
    EventFlavor[EventFlavor["ScrollDown"] = 3] = "ScrollDown";
    EventFlavor[EventFlavor["ScrollLeft"] = 4] = "ScrollLeft";
    EventFlavor[EventFlavor["ScrollRight"] = 5] = "ScrollRight";
    EventFlavor[EventFlavor["SwipeUp"] = 6] = "SwipeUp";
    EventFlavor[EventFlavor["SwipeDown"] = 7] = "SwipeDown";
    EventFlavor[EventFlavor["SwipeLeft"] = 8] = "SwipeLeft";
    EventFlavor[EventFlavor["SwipeRight"] = 9] = "SwipeRight";
    EventFlavor[EventFlavor["PressShort"] = 10] = "PressShort";
    EventFlavor[EventFlavor["PressLong"] = 11] = "PressLong";
})(EventFlavor || (EventFlavor = {}));
export var EventMode;
(function (EventMode) {
    EventMode[EventMode["KeyUp"] = 0] = "KeyUp";
    EventMode[EventMode["KeyDown"] = 1] = "KeyDown";
    EventMode[EventMode["ScrollUp"] = 2] = "ScrollUp";
    EventMode[EventMode["ScrollDown"] = 3] = "ScrollDown";
    EventMode[EventMode["ScrollLeft"] = 4] = "ScrollLeft";
    EventMode[EventMode["ScrollRight"] = 5] = "ScrollRight";
    EventMode[EventMode["SwipeUp"] = 6] = "SwipeUp";
    EventMode[EventMode["SwipeDown"] = 7] = "SwipeDown";
    EventMode[EventMode["SwipeLeft"] = 8] = "SwipeLeft";
    EventMode[EventMode["SwipeRight"] = 9] = "SwipeRight";
    EventMode[EventMode["PressShort"] = 10] = "PressShort";
    EventMode[EventMode["PressLong"] = 11] = "PressLong";
})(EventMode || (EventMode = {}));
